trie/database_wrapper.go

```diff
@@ -0,0 +1,245 @@
+	}
+	return db
+}
+
+// GetReader returns a reader for accessing all trie nodes with provided
+// state root. Nil is returned in case the state is not available.
+func (db *Database) GetReader(blockRoot common.Hash) Reader {
+	return db.backend.GetReader(blockRoot)
+}
+
+// Update performs a state transition by committing dirty nodes contained
+// in the given set in order to update state from the specified parent to
+// the specified root. The held pre-images accumulated up to this point
+// will be flushed in case the size exceeds the threshold.
+func (db *Database) Update(root common.Hash, parent common.Hash, nodes *MergedNodeSet) error {
+	if db.preimages != nil {
+		db.preimages.commit(false)
+	}
+	return db.backend.Update(root, parent, nodes)
+}
+
+// Commit iterates over all the children of a particular node, writes them out
+// to disk. As a side effect, all pre-images accumulated up to this point are
+// also written.
+func (db *Database) Commit(root common.Hash, report bool) error {
+	if db.preimages != nil {
+		db.preimages.commit(true)
+	}
+	return db.backend.Commit(root, report)
+}
+
+// Size returns the storage size of dirty trie nodes in front of the persistent
+// database and the size of cached preimages.
+func (db *Database) Size() (common.StorageSize, common.StorageSize) {
+	var (
+		storages  common.StorageSize
+		preimages common.StorageSize
+	)
+	storages = db.backend.Size()
+	if db.preimages != nil {
+		preimages = db.preimages.size()
+	}
+	return storages, preimages
+}
+
+// Initialized returns an indicator if the state data is already initialized
+// according to the state scheme.
+func (db *Database) Initialized(genesisRoot common.Hash) bool {
+	return db.backend.Initialized(genesisRoot)
+}
+
+// Scheme returns the node scheme used in the database.
+func (db *Database) Scheme() string {
+	return db.backend.Scheme()
+}
+
+// Close flushes the dangling preimages to disk and closes the trie database.
+// It is meant to be called when closing the blockchain object, so that all
+// resources held can be released correctly.
+func (db *Database) Close() error {
+	if db.preimages != nil {
+		db.preimages.commit(true)
+	}
+	return db.backend.Close()
+}
+
+// saveCache saves clean state cache to given directory path
+// using specified CPU cores.
+func (db *Database) saveCache(dir string, threads int) error {
+	if db.cleans == nil {
+		return nil
+	}
+	log.Info("Writing clean trie cache to disk", "path", dir, "threads", threads)
+
+	start := time.Now()
+	err := db.cleans.SaveToFileConcurrent(dir, threads)
+	if err != nil {
+		log.Error("Failed to persist clean trie cache", "error", err)
+		return err
+	}
+	log.Info("Persisted the clean trie cache", "path", dir, "elapsed", common.PrettyDuration(time.Since(start)))
+	return nil
+}
+
+// SaveCache atomically saves fast cache data to the given dir using all
+// available CPU cores.
+func (db *Database) SaveCache(dir string) error {
+	return db.saveCache(dir, runtime.GOMAXPROCS(0))
```

Comment:
`runtime.GOMAXPROCS(0)` will not change the setting, it's used to retrieve the previous setting(e.g. how many CPU cores we can use).

`runtime.NumCPU` is not used since it returns the number of cpu cores, but we may not be permitted to use that many.